deploy/buildtools/buildfarm.py

```diff
@@ -0,0 +1,334 @@
+import logging
+import sys
+import abc
+
+from awstools.awstools import *
+
+# imports needed for python type checking
+from typing import cast, Any, Dict, Optional, Sequence, List, TYPE_CHECKING
+from mypy_boto3_ec2.service_resource import Instance as EC2InstanceResource
+# needed to avoid type-hint circular dependencies
+# TODO: Solved in 3.7.+ by "from __future__ import annotations" (see https://stackoverflow.com/questions/33837918/type-hints-solve-circular-dependency)
+#       and normal "import <module> as ..." syntax (see https://www.reddit.com/r/Python/comments/cug90e/how_to_not_create_circular_dependencies_when/)
+if TYPE_CHECKING:
+    from buildtools.buildconfig import BuildConfig
+else:
+    BuildConfig = object
+
+rootLogger = logging.getLogger()
+
+class BuildHost:
+    """Abstract class representing a single build host which holds a single build config.
+
+    Attributes:
+        build_config: Build config associated with the build host.
+        dest_build_dir: Name of build dir on build host.
+        ip_address: IP address of build host.
+    """
+    build_config: BuildConfig
+    dest_build_dir: str
+    ip_address: str
+
+    def __init__(self):
+        self.ip_address = ""
+        self.build_config = None # type: ignore
```

Comment:
Sounds good. I wanted to avoid creating a bunch of `Optional`s everywhere. Depending on how the `parse_args` inlining looks I'll make some modifications (I tend to like the assertion solution better).